### ISSUE TITLE
HLS Looping

### ISSUE THREAD

**Issue by webmonch** [superpoweredSDK/Low-Latency-Android-iOS-Linux-Windows-tvOS-macOS-Interactive-Audio-Platform#777]
Hi,

is it possible to loop HLS file? Can't seem to get it working

**Comment by ivannador** [comment]
Hello,

No, it is not possible to loop a HLS file.

**Comment by webmonch** [comment]
Thanks for a quick response @ivannador !

What about progressive download?

**Comment by gaborszanto** [comment]
A progressive download will have all features, including looping.

**Comment by webmonch** [comment]
Thanks @gaborszanto!

And how does **cachePosition()** behave with progressive download? Can it be called right after **open(...)**?

**Comment by gaborszanto** [comment]
It gets cached when that point is downloaded.

**Comment by webmonch** [comment]
Thanks! Really appreciate quick response 🙏

One (hopefully) last question - I am trying to run ReactNative example (with progressive download) and I am getting an error:

`PlayerEvent_ConnectionLost: "Network socket error. No internet?"`

Both on simulator and on device. 

If I change it to HLS streaming - everything works fine.

Song url is from example: https://docs.superpowered.com/audio/samples/splice/SO_PF_74_string_phrase_soaring_Gb.mp3
XCode 15.3 
IOS 17.4

All settings are from example project.

**Comment by webmonch** [comment]
Same on Android. HLS plays fine, progressive download doesn't work.

**Comment by webmonch** [comment]
@ivannador @gaborszanto

**Comment by ivannador** [comment]
Did you try it with other examples? Did you try a different URL provided by you?

**Comment by webmonch** [comment]
@ivannador  Yes, this one works. But the one in the example doesn't, though it is a valid url. 
Is there a way to tell why? To make sure it doesn't happen with urls provided buy us.

Also when I open working url, I get **Superpowered::AdvancedAudioPlayer::PlayerEvent_Opened** right before **Superpowered::AdvancedAudioPlayer::PlayerEvent_ProgressiveDownloadFinished**,  so playback starts only after download finished. 

I expected progressive download to start while the track is being downloaded, or is this not the case?

**Comment by ivannador** [comment]
Your server has to handle `Content-Range` in the response header and `Range` in the request header. Make sure it does, then it'll work. For example S3 supports it by default.
This not working with the URL from the example is an issue on our server, we'll fix that.

Well it is ready for playback after `PlayerEvent_Opened` so you should listen for that. If `PlayerEvent_ProgressiveDownloadFinished` comes quickly it means it is downloaded or it's already downloaded and loaded from local.

**Comment by webmonch** [comment]
Thank you.

I am also getting a small click when using **loopBetween(3000, 5000, true, 255, false)** as you can hear on the video. Is it possible to remove it completely?
Happens both on iOS (device + simulator) and MacOS.


https://github.com/superpoweredSDK/Low-Latency-Android-iOS-Linux-Windows-tvOS-macOS-Interactive-Audio-Platform/assets/1910471/1a1af622-1062-43e2-8003-c94d02018789

Here is the file I am using (local playback, not download)
https://test-reports-frontend-web.s3.eu-west-2.amazonaws.com/track.mp3

**Comment by webmonch** [comment]
And regarding seeking in file that is being downloaded, from what I gathered if you seek to the part that hasn't yet been downloaded, nothing plays, is that correct? If yes, is it possible to force download start from a new position?

**Comment by webmonch** [comment]
@ivannador is it technically possible using buffers and advanced player to implement progressive download with seeking? 

My use case is basically this: 
- there is an audio file that is split into logical sections. The start/end/length in ms of each section is known before payback starts, along with file size in bytes.
- when user plays this file from url, file download and playback starts (like now with progressive download)
- user can Seek() to the start of another section of this file. Playback should start from the beginning of that section. if that part of the file hasn't yet been downloaded - download must start immediately without waiting the progress to reach it (kinda like HLS player does). 
- sections can be looped seamlessly with loopBetween.

From what I gathered, I basically need an HLS player with looping or a progressive download that supports seeking to not downloaded parts.

How can I implement this? 

If this is not possible with built-in features, I can implement file downloading in chunks with range header, etc., 
But how would I integrate this with audio buffers?

**Comment by ivannador** [comment]
> Thank you.
> 
> I am also getting a small click when using **loopBetween(3000, 5000, true, 255, false)** as you can hear on the video. Is it possible to remove it completely? Happens both on iOS (device + simulator) and MacOS.
> RPReplay_Final1714057324.mov
> 
> Here is the file I am using (local playback, not download) https://test-reports-frontend-web.s3.eu-west-2.amazonaws.com/track.mp3

We currently don't have any windowing functionality when looping. Using a sine wave the DC offset is very prominent when looping, not so much when using more complex waveforms like actual music. You can implement some level of fade-in/fade-out in the audio processing loop by getting the current position and ducking near the loop jump.

**Comment by ivannador** [comment]
Progressive download works in a linear fashion. It starts to download the audio data continuously, you can only seek to what is already downloaded. I suppose making it possible on server-side to "cut" the track in specific parts and load those chunks with progressive download is possible.

That click in `loopBetween` seems to be a bug, good catch. It is very prominent with a pure tone you are using, not so much with a more complex waveform like an actual track, but still.
We'll provide a fix for that.

As for separately downloading chunks, you can use the `AudioInMemory` class with append, and `openMemory` in Advanced Audio Player.

**Comment by webmonch** [comment]
Could you please describe very high level flow of using AudioInMemory with the mentioned use case?

e.g. when I download first 10% of the file and 10% in the middle of the file.
**decodeToAudioInMemory** per docs decodes entire file. How to I decode non linear chunks and play them with the option to loopBetween?

**Comment by ivannador** [comment]
Well you can create an `AudioInMemory` container and append raw PCM data to it while you already opened it with `openMemory` of AAP.
I think you have to make sure that on the server you make downloadable chunks out of the audio track to decode it with Decoder non-linearly.

**Comment by webmonch** [comment]
Thank you,

Do you think it will possible to add looping to HLS audio file streaming (not real-time, just aac converted to hls) after we purchase a license?

**Comment by webmonch** [comment]
Also do I need to call any method to inform the player that new data has been appended (when appending compressed data into AiM buffers)?

**Comment by ivannador** [comment]
We have HLS improvements planned on the longer roadmap, but I can't tell when it will come.

You don't have to inform the player that data has been appended.

**Comment by webmonch** [comment]
For some reason player doesn't see new sections. 
I download chunks of HLS every few seconds and append compressed audio to AiM, but player plays only those that were added before openMemory() was called.

What I do:
- first two 3-second chunks (aac hls) are downloaded and appended to Audio in Memory 
- **player->openMemory(compressedAudioBuffer)**
- wait for **Opened** event
- continue downloading next 2 chunks and appending them
- **player->play()**

The **reserved** field in compressed audio buffer changes so I guess player noticed the chunks but the player plays only first  two chunks. 
What am I missing here?

Screenshot A - when I call play() first time (plays 2 chunks, though 4 are in buffers)
Screenshot B - when I call play() again after first play (still plays only 2)

This is compressed buffer so size is in bytes.

<img width="401" alt="Screenshot 2024-05-03 at 12 30 18" src="https://github.com/superpoweredSDK/Low-Latency-Android-iOS-Linux-Windows-tvOS-macOS-Interactive-Audio-Platform/assets/1910471/bf01e0ca-20d4-4012-8706-7fdc395cbcc3">

<img width="402" alt="Screenshot 2024-05-03 at 12 30 35" src="https://github.com/superpoweredSDK/Low-Latency-Android-iOS-Linux-Windows-tvOS-macOS-Interactive-Audio-Platform/assets/1910471/acc57081-a983-47e4-8ff6-0c19c755ff02">

**Comment by gaborszanto** [comment]
Is the size argument for the AudioInMemory (main table) set to 0?

**Comment by ivannador** [comment]
Hey @webmonch,

Did you manage to test this with our latest release (2.6.8)?

**Comment by webmonch** [comment]
@ivannador yes but still had issues. Here's the code I am using (based on HLS streaming project for IOS from your examples)

https://gist.github.com/webmonch/5cce8ad2a612941e4e327db89edcea7f

Wrote to your zendesk about it yesterday.
